crypto/curve25519/curve25519.c

```diff
@@ -31,33 +31,223 @@
 #include "../internal.h"
 #include "../fipsmodule/cpucap/internal.h"
 
-// x25519_NEON is defined in asm/x25519-arm.S.
 #if defined(OPENSSL_ARM) && !defined(OPENSSL_NO_ASM) && !defined(OPENSSL_APPLE)
-#define BORINGSSL_X25519_NEON
+#define BORINGSSL_X25519_NEON_CAPABLE
 #endif
 
+#if (defined(OPENSSL_X86_64) || defined(OPENSSL_AARCH64)) && \
+    (defined(OPENSSL_LINUX) || defined(OPENSSL_APPLE)) && \
+    !defined(OPENSSL_NO_ASM) && !defined(MY_ASSEMBLER_IS_TOO_OLD_FOR_AVX)
+#include "../../third_party/s2n-bignum/include/s2n-bignum_aws-lc.h"
+#define CURVE25519_S2N_BIGNUM_CAPABLE
+#endif
 
-void x25519_NEON(uint8_t out[32], const uint8_t scalar[32],
-                 const uint8_t point[32]);
 
-#if defined(BORINGSSL_X25519_NEON)
+OPENSSL_INLINE int x25519_s2n_bignum_capable(void) {
+#if defined(CURVE25519_S2N_BIGNUM_CAPABLE)
+  return 1;
+#else
+  return 0;
+#endif
+}
 
-OPENSSL_INLINE int curve25519_asm_capable(void) {
+OPENSSL_INLINE int x25519_Armv7_neon_capable(void) {
+#if defined(BORINGSSL_X25519_NEON)
   return CRYPTO_is_NEON_capable();
+#else
+  return 0;
+#endif
 }
 
-#else
+// Stub functions if implementations are not compiled.
 
-OPENSSL_INLINE int curve25519_asm_capable(void) {
-  return 0;
+#if !defined(CURVE25519_S2N_BIGNUM_CAPABLE)
+
+void curve25519_x25519_byte(uint8_t res[32], const uint8_t scalar[32],
+  const uint8_t point[32]);
+void curve25519_x25519_byte_alt(uint8_t res[32], const uint8_t scalar[32],
+  const uint8_t point[32]);
+void curve25519_x25519base_byte(uint8_t res[32], const uint8_t scalar[32]);
+void curve25519_x25519base_byte_alt(uint8_t res[32], const uint8_t scalar[32]);
+
+void curve25519_x25519_byte(uint8_t res[32], const uint8_t scalar[32],
+  const uint8_t point[32]) {
+  abort();
+}
```

Comment:
Can we avoid these aborts? How did P384/521 manage this case? Can we gracefully fallback to the existing implementation and not call the s2n-bignum code if the runtime machine doesn't support it?

Reply:
This code path is only taken if:
* architecture support is signaled at compile-time (i.e. x86_64 or aarch64).
* run-time checks pass

It's is equivalent to the p{384,521} implementation. Except, this implementation is more robust, since the p{384,521} cannot fall-back to the C-implementation at run-time: decision to use s2n-bignum is solely done at compile-time (AFAICT). But we do this in `X25519()`.

If both implementations somehow tries to use an s2n-bignum implementation when not supported (in this case of this PR this would be a bug in the logic), both this PR and p{384,521} will have the same failure mode: either abort of fail with an invalid instruction.

Reply:
The logic to fall-back to a different implementation is implemented in e.g. `X25519()`.